lms/security.py

```diff
@@ -42,15 +42,58 @@ class Permissions(Enum):
 @lru_cache(maxsize=1)
 def get_policy(request):
     """Pick the right policy based the request's path."""
+    # pylint:disable=too-many-return-statements
```

Comment:
This method is not very sophisticated. The aim here is to get the explicitlness in place first, we can refine later.

We could move this direclty to the `add_route` side of things with maybe a route predicate https://docs.pylonsproject.org/projects/pyramid/en/latest/api/config.html#pyramid.config.Configurator.add_route_predicate.

Reply:
:+1: I think it's fine to disable the Pylint warning here.

If you wanted to get rid of the warning, I think what Pylint is trying to tell you to do is to separate configuration from logic. What you have here is some logic that takes as its input a list of `(boolean_function, security_policy_factory)` 2-tuples. It iterates over the list calling each boolean function and if the boolean function returns `True` then it uses that security policy factory.

You probably wouldn't want to do it quite like this (with a bunch of tuples and lambdas) but here's an example implementation that has no Pylint warnings and still passes all the tests.

I think a better version of this would probably recognise that all of the boolean functions are pattern-matching against `path` (they're all `path in <list_of_paths>` or `path.startswith(prefix)` or `path.endswith(suffix)`). This can be implemented by associating a regular expression or list of regular expressions with each security policy factory, rather than a boolean function.

```python
@lru_cache(maxsize=1)
def get_policy(request):
    """Pick the right policy based the request's path."""
    path = request.path

    policies = [
        (
            lambda: path.startswith("/admin") or path.startswith("/googleauth"),
            LMSGoogleSecurityPolicy,
        ),
        (
            lambda: path
            in {"/lti_launches", "/content_item_selection", "/api/gateway/h/lti"},
            lambda: LTIUserSecurityPolicy(get_lti_user_from_launch_params),
        ),
        (
            lambda: path
            in {
                "/canvas_oauth_callback",
                "/api/blackboard/oauth/callback",
                "/api/d2l/oauth/callback",
            },
            lambda: LTIUserSecurityPolicy(get_lti_user_from_oauth_callback),
        ),
        (
            lambda: path.startswith("/api") and path.endswith("authorize"),
            lambda: LTIUserSecurityPolicy(
                partial(get_lti_user_from_bearer_token, location="querystring")
            ),
        ),
        (
            lambda: path.startswith("/api")
            or path
            in {
                "/lti/1.3/deep_linking/form_fields",
                "/lti/1.1/deep_linking/form_fields",
            },
            lambda: LTIUserSecurityPolicy(
                partial(get_lti_user_from_bearer_token, location="headers")
            ),
        ),
        (
            lambda: path in {"/assignment"},
            lambda: LTIUserSecurityPolicy(
                partial(get_lti_user_from_bearer_token, location="form")
            ),
        ),
    ]

    for boolean_func, securitypolicy_factory in policies:
        if boolean_func():
            return securitypolicy_factory()

    return UnAutheticatedSecurityPolicy()
```

Reply:
Might wanna just take `request.path` as the argument in the first place (`get_policy(path)`) since `request.path` is the only thing that `get_policy()` uses `request` for

Reply:
`get_policy()` could probably be a `@staticmethod` within `SecurityPolicy` since that's the only class that uses it

Reply:
Kept the method silly for now. I'd like to do another pass that just focuses in making it nicer but once that the policy <-> path is already merged.